### PR TITLE
Rewrite `zap_srcref()` in C

### LINKED ISSUE #1513: Reimplement `zap_srcref()` in C
It's causing a bottleneck in large shiny apps.

### PR DESCRIPTION
**Pull request by lionel-**
Closes #1513.

@wch could you try this branch out and see if it fixes your performance issues please?

### CODE REVIEW

File: src/rlang/node.c [hunk 1]
```diff
@@ -30,7 +30,9 @@ r_obj* r_new_pairlist(const struct r_pair* args,
 
 // Shallow copy of a node tree
 r_obj* r_node_tree_clone(r_obj* x) {
```
Comment (lionel-) [inline src/rlang/node.c:32]: I didn't end up using this, but the generalisation to both calls and pairlists makes sense.

**Review by DavisVaughan — APPROVED** [review]

**Comment by wch** [comment]
This does indeed speed up the startup time for shiny app I was testing before.

A simple benchmark without shiny:

```R
library(rlang)
system.time({
  for (i in 1:1000) {
    zap_srcref(lm)
  }
})
```

Results with CRAN version:

```
   user  system elapsed 
  0.891   0.015   0.910 
```

Results with this PR:

```
   user  system elapsed 
  0.017   0.001   0.017 
```

File: src/internal/attr.c [hunk 2]
```diff
@@ -194,6 +194,66 @@ r_ssize length_dispatch(r_obj* x, r_obj* env) {
...
+
+static
+r_obj* call_zap_srcref(r_obj* x) {
+  x = KEEP(r_clone(x));
+
+  attrib_zap_srcref(x);
+
+  if (r_node_car(x) == r_syms.function) {
+    r_node_poke_cdr(r_node_cddr(x), r_null);
+  }
```
[the remark this answers is not in the document]
Comment (lionel-) [reply src/internal/attr.c:225]: Good point, it now says:

```c
    // Remove `call[[4]]` where the parser stores srcref information
    // for calls to `function`
```

File: src/internal/attr.c [hunk 3]
```diff
@@ -194,6 +194,66 @@ r_ssize length_dispatch(r_obj* x, r_obj* env) {
...
+  default: return x;
+  }
+}
+
+static
+r_obj* fn_zap_srcref(r_obj* x) {
+  x = KEEP(r_clone(x));
+
+  r_fn_poke_body(x, zap_srcref(r_fn_body(x)));
+  r_attrib_poke(x, r_syms.srcref, r_null);
```
[the remark this answers is not in the document]
Comment (lionel-) [reply src/internal/attr.c:211]: Not to my knowledge

**Pull request merged**